qtfred/src/mission/dialogs/ShipEditor/ShipCustomWarpDialogModel.cpp

```diff
@@ -0,0 +1,280 @@
+#include "ShipCustomWarpDialogModel.h"
+
+#include "ship/shipfx.h"
+namespace fso::fred::dialogs {
+ShipCustomWarpDialogModel::ShipCustomWarpDialogModel(QObject* parent, EditorViewport* viewport, bool departure)
+	: AbstractDialogModel(parent, viewport), _m_departure(departure)
+{
+	initializeData();
+}
+
+bool ShipCustomWarpDialogModel::apply()
+{
+	WarpParams params;
+	params.direction = _m_departure ? WarpDirection::WARP_OUT : WarpDirection::WARP_IN;
+
+	if (_m_warp_type < Num_warp_types) {
+		params.warp_type = _m_warp_type;
+	} else {
+		params.warp_type = (_m_warp_type - Num_warp_types) | WT_DEFAULT_WITH_FIREBALL;
+	}
+
+	if (!_m_start_sound.empty()) {
+		gamesnd_id id = gamesnd_get_by_name(_m_start_sound.c_str());
+		if (id.value() == -1) {
+			Warning(LOCATION, "Game Sound \"%s\" does not exist. Skipping", _m_start_sound.c_str());
+		} else {
+			params.snd_start = id;
+		}
+	}
+
+	if (!_m_end_sound.empty()) {
+		gamesnd_id id = gamesnd_get_by_name(_m_end_sound.c_str());
+		if (id.value() == -1) {
+			Warning(LOCATION, "Game Sound \"%s\" does not exist. Skipping", _m_end_sound.c_str());
+		} else {
+			params.snd_end = id;
+		}
+	}
+
+	if (_m_departure && _m_warpout_engage_time) {
+		params.warpout_engage_time = fl2i(_m_warpout_engage_time * 1000.0f);
+	}
+	if (_m_speed) {
+		params.speed = _m_speed;
+	}
+	if (_m_time) {
+		params.time = fl2i(_m_time * 1000.0f);
+	}
+	if (_m_accel_exp) {
+		params.accel_exp = _m_accel_exp;
+	}
+	if (_m_radius) {
+		params.accel_exp = _m_radius;
+	}
+	if (!_m_anim.empty()) {
+		strcpy_s(params.anim, _m_anim.c_str());
+	}
+	params.supercap_warp_physics = _m_supercap_warp_physics;
+	if (_m_departure && _m_player_warpout_speed) {
+		params.warpout_player_speed = _m_player_warpout_speed;
+	}
+	int index = find_or_add_warp_params(params);
+
+	for (object* objp = GET_FIRST(&obj_used_list); objp != END_OF_LIST(&obj_used_list); objp = GET_NEXT(objp)) {
```

Comment:
You can use `for (object* objp: list_range(&obj_used_list))` here and in similar loops if you want.